api/applications.go

```diff
@@ -0,0 +1,168 @@
+	ListApplications(ctx context.Context, filter string) ([]models.Applicationable, error)
+	AddApplicationPassword(ctx context.Context, applicationObjectID string, displayName string, endDateTime time.Time) (models.PasswordCredentialable, error)
+	RemoveApplicationPassword(ctx context.Context, applicationObjectID string, keyID *uuid.UUID) error
+}
+
+var _ ApplicationsClient = (*AppClient)(nil)
+var _ GroupsClient = (*AppClient)(nil)
+var _ ServicePrincipalClient = (*AppClient)(nil)
+
+type AppClient struct {
+	client *msgraphsdkgo.GraphServiceClient
+}
+
+// Reference: https://docs.microsoft.com/en-us/graph/deployments#microsoft-graph-and-graph-explorer-service-root-endpoints
+func GetGraphURI(env string) (string, error) {
+	switch env {
+	case "AzurePublicCloud", "":
+		return "https://graph.microsoft.com", nil
+	case "AzureUSGovernmentCloud":
+		return "https://graph.microsoft.us", nil
+	case "AzureGermanCloud":
+		return "https://graph.microsoft.de", nil
+	case "AzureChinaCloud":
+		return "https://microsoftgraph.chinacloudapi.cn", nil
+	default:
+		return "", fmt.Errorf("environment '%s' unknown", env)
+	}
+}
+
+// NewMSGraphApplicationClient returns a new AppClient configured to interact with
+// the Microsoft Graph API. It can be configured to target alternative national cloud
+// deployments via graphURI. For details on the client configuration see
+// https://learn.microsoft.com/en-us/graph/sdks/national-clouds
+func NewMSGraphApplicationClient(graphURI string, creds azcore.TokenCredential) (*AppClient, error) {
+	scopes := []string{
+		fmt.Sprintf("%s/.default", graphURI),
+	}
+
+	authProvider, err := auth.NewAzureIdentityAuthenticationProviderWithScopes(creds, scopes)
+	if err != nil {
+		return nil, err
+	}
+
+	adapter, err := msgraphsdkgo.NewGraphRequestAdapter(authProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	adapter.SetBaseUrl(fmt.Sprintf("%s/v1.0", graphURI))
+	client := msgraphsdkgo.NewGraphServiceClient(adapter)
+
+	ac := &AppClient{
+		client: client,
+	}
+	return ac, nil
+}
+
+func (c *AppClient) GetApplication(ctx context.Context, clientID string) (models.Applicationable, error) {
+	filter := fmt.Sprintf("appId eq '%s'", clientID)
+	req := applications.ApplicationsRequestBuilderGetRequestConfiguration{
+		QueryParameters: &applications.ApplicationsRequestBuilderGetQueryParameters{
+			Filter: &filter,
+		},
+	}
+
+	resp, err := c.client.Applications().Get(ctx, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	apps := resp.GetValue()
+	if len(apps) == 0 {
+		return nil, fmt.Errorf("no application found")
+	}
+	if len(apps) > 1 {
+		return nil, fmt.Errorf("multiple applications found - double check your client_id")
+	}
+
+	return apps[0], nil
+}
+
+func (c *AppClient) ListApplications(ctx context.Context, filter string) ([]models.Applicationable, error) {
+	headers := abstractions.NewRequestHeaders()
+	headers.Add("ConsistencyLevel", "eventual")
```

Comment:
Doesn't look like we actually need it? Since we are only ever using the "eq" filter? https://learn.microsoft.com/en-us/graph/aad-advanced-queries?tabs=go

Reply:
Ah interesting, thanks for sending that link for reference! I think we might be okay to not use the header, I was mainly going off of the official [MSGraph API documentation](https://learn.microsoft.com/en-us/graph/api/application-list?view=graph-rest-1.0&tabs=go#example-3-use-filter-and-top-to-get-one-application-with-a-display-name-that-starts-with-a-including-a-count-of-returned-objects), where they seem to be setting the header for the List Applications call even though they are using a filter 🤔 

I'll test without the header and if it works I'll remove it 👍🏼

Reply:
Ah, now that I read the docs more closely, it _might_ be that we only need that header if we use orderby. Y'all are right, looks like we don't need the header! Thanks for the catch 🙏🏼